instant-xml-macros/src/de.rs

```diff
@@ -0,0 +1,266 @@
+                    __Ignore,
+                }
+
+                fn get_attribute(value: &str) -> __Attributes {
+                    #attributes_consts
+                    match value {
+                        #attributes_names
+                        _ => __Attributes::__Ignore
+                    }
+                }
+
+                struct StructVisitor;
+                impl<'xml> Visitor<'xml> for StructVisitor {
+                    type Value = #ident;
+
+                    fn visit_struct<'a>(&self, deserializer: &mut ::instant_xml::Deserializer) -> Result<Self::Value, ::instant_xml::Error>
+                    {
+                        #declare_values
+                        println!("visit struct");
+
+                        while let Some(( key, _ )) = deserializer.peek_next_attribute() {
+                            match get_attribute(&key) {
+                                #attr_type_match
+                                __Attributes::__Ignore => todo!(),
+                            }
+                        }
+                        while let Some(item) = &deserializer.peek_next_tag()? {
+                            match item {
+                                XmlRecord::Open(item) => {
+                                    match get_element(&item.key.as_ref()) {
+                                        #elem_type_match
+                                        __Elements::__Ignore => todo!(),
+                                    }
+                                 }
+                                 XmlRecord::Close(tag) => {
+                                    println!("Close: {}", tag);
+                                    if tag == &#name {
+                                        break;
+                                    }
+                                },
+                                XmlRecord::Element(_) => panic!("Unexpected element"),
+                            }
+                        }
+
+                        println!("return");
+                        Ok(Self::Value {
+                            #return_val
+                    })
+                    }
+                }
+
+                #namespaces_map;
+                deserializer.deserialize_struct(StructVisitor{}, #name, &namespaces_map)
+            }
+        ));
+
+        out.extend(quote!(
+                const TAG_NAME: ::instant_xml::XMLTagName<'xml> = ::instant_xml::XMLTagName::Custom(#name);
+            )
+        );
+
+        Deserializer { out }
+    }
+
+    fn process_field(
+        field: &syn::Field,
+        index: usize,
+        declare_values: &mut TokenStream,
+        return_val: &mut TokenStream,
+        tokens: &mut Tokens,
+        is_element: bool,
+    ) {
+        let field_name = field.ident.as_ref().unwrap().to_string();
+        let const_field_name = Ident::new(&field_name.to_uppercase(), Span::call_site());
+        let field_value = field.ident.as_ref().unwrap();
```

Comment:
We probably don't need `field_value` since we already have `field_name`, which amounts to the same thing? Should factor out the common `.as_ref().unwrap()` at least.

Reply:
We use both of them in `quote!`, so I believe we need both here, unless there is other way to pass it?

Reply:
I just mean we should factor out the common part. And the naming seems unclear given that both have the same value?

Reply:
Changed to field_var and field_var_str, if you have some better naming idea, please let me know.